snuba/clickhouse/optimize/optimize.py

```diff
@@ -286,48 +289,48 @@ def optimize_partition_runner(
     1. There are no more partitions which need optimization.
     2. The final cutoff time is reached. In this case, the scheduler will
     raise an exception which would be propagated to the caller.
+
+    Details of execution flow:
+    1. start by reading configured_num_threads from the Snuba Admin runtime config
+    2. dispatches configured_num_threads threads to optimize configured_num_threads partitions (1 partition per thread)
+    3. as soon as one thread finishes, check configured_num_threads from runtime config again
+    4. if configured_num_threads > number of currently active threads, dispatch more threads
     """
-    remaining_partitions = partitions
-    while remaining_partitions:
-        schedule = scheduler.get_next_schedule(remaining_partitions)
-        num_threads = len(schedule.partitions)
-        logger.info(
-            f"Running schedule with cutoff time: "
-            f"{schedule.cutoff_time} with {num_threads} threads"
-        )
-        threads: MutableSequence[threading.Thread] = []
-        for i in range(0, num_threads):
-            threads.append(
-                threading.Thread(
-                    target=optimize_partitions,
-                    args=(
+    scheduler = OptimizeScheduler(default_parallel_threads=default_parallel_threads)
+
+    with ThreadPoolExecutor(max_workers=32) as executor:
+        pending_futures: set[Future[Any]] = set()
+
+        partitions_to_optimize = deque(partitions)
+        while partitions_to_optimize:
```

Comment:
I wonder if we can simplify two loops into just one:
```
pending_futures: set[Future[Any]] = set()
partitions_to_optimize = deque(partitions)
configured_num_threads = get_num_threads(default_parallel_threads)
while partitions_to_optimize and len(pending_futures) < configured_num_threads:
    schedule = scheduler.get_next_schedule(partitions_to_optimize)
    logger.info(...)
    pending_futures.add(...)
    completed_futures, pending_futures = wait(pending_futures,return_when=FIRST_COMPLETED)
    
    for future in completed_futures:
        future.result()
```

In fact, I think the two conditions in the `while` loop are testing for the same thing, so we should be able to even simplify it further to just one condition like `while partitions_to_optimize`:

Reply:
If we simplify it to a single loop like that, then you are only checking the configured number of threads once, before the `while` loop. We want to adjust the number of threads on the fly.

> I think the two conditions in the while loop are testing for the same thing, so we should be able to even simplify it further to just one condition like while partitions_to_optimize

if we take out `len(pending_futures) < configured_num_threads`, then we can potentially schedule more partitions to optimize than the number of threads. The goal is to never schedule more partitions than the number of configured threads